File: DiKErnel/Test/DiKErnel.FunctionLibrary.Test/AsphaltRevetmentWaveImpact/AsphaltRevetmentWaveImpactFunctionsInputTest.cpp
// Copyright (C) Stichting Deltares 2022. All rights reserved.
//
// This file is part of DiKErnel.
//
// DiKErnel is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, version 3.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.
//
// All names, logos, and references to "Deltares" are registered trademarks of
// Stichting Deltares and remain full property of Stichting Deltares at all times.
// All rights reserved.

#include <gtest/gtest.h>

#include "AsphaltRevetmentWaveImpactFunctionsInput.h"

namespace DiKErnel::FunctionLibrary::Test
{
    using namespace std;

    TEST(AsphaltRevetmentWaveImpactFunctionsInputTest, GivenAsphaltRevetmentWaveImpactFunctionsInput_WhenCreated_ThenExpectedValues)
    {
        // When
        const AsphaltRevetmentWaveImpactFunctionsInput input{};

        // Then
        ASSERT_DOUBLE_EQ(0.0, input._logFailureTension);
        ASSERT_DOUBLE_EQ(0.0, input._averageNumberOfWaves);
        ASSERT_DOUBLE_EQ(0.0, input._maximumPeakStress);
        ASSERT_DOUBLE_EQ(0.0, input._stiffnessRelation);
        ASSERT_DOUBLE_EQ(0.0, input._computationalThickness);
        ASSERT_DOUBLE_EQ(0.0, input._outerSlope);

        const auto factors = vector<pair<double, double>>();
        ASSERT_EQ(factors, input._widthFactors);
        ASSERT_EQ(factors, input._depthFactors);
        ASSERT_EQ(factors, input._impactFactors);

        ASSERT_DOUBLE_EQ(0.0, input._z);
        ASSERT_DOUBLE_EQ(0.0, input._waterLevel);
        ASSERT_DOUBLE_EQ(0.0, input._waveHeightHm0);
        ASSERT_DOUBLE_EQ(0.0, input._fatigueAlpha);
        ASSERT_DOUBLE_EQ(0.0, input._fatigueBeta);
        ASSERT_DOUBLE_EQ(0.0, input._impactNumberC);
    }
}

File: Test/DiKErnel.FunctionLibrary.Test/AsphaltRevetmentWaveImpact/AsphaltRevetmentWaveImpactFunctionsInputTest.cs
// Copyright (C) Stichting Deltares and State of the Netherlands 2023. All rights reserved.
//
// This file is part of DiKErnel.
//
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 3 of the License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with this program; if not, see <https://www.gnu.org/licenses/>.
//
// All names, logos, and references to "Deltares" are registered trademarks of
// Stichting Deltares and remain full property of Stichting Deltares at all times.
// All rights reserved.

using System;
using System.Collections.Generic;
using DiKErnel.FunctionLibrary.AsphaltRevetmentWaveImpact;
using NUnit.Framework;
using Random = DiKErnel.TestUtil.Random;

namespace DiKErnel.FunctionLibrary.Test.AsphaltRevetmentWaveImpact
{
    [TestFixture]
    public class AsphaltRevetmentWaveImpactFunctionsInputTest
    {
        [Test]
        public void Constructor_ExpectedValues()
        {
            // Setup
            double logFailureTension = Random.NextDouble();
            double averageNumberOfWaves = Random.NextDouble();
            double maximumPeakStress = Random.NextDouble();
            double stiffnessRelation = Random.NextDouble();
            double computationalThickness = Random.NextDouble();
            double outerSlope = Random.NextDouble();

            IReadOnlyList<(double, double)> widthFactors = Array.Empty<(double, double)>();
            IReadOnlyList<(double, double)> depthFactors = Array.Empty<(double, double)>();
            IReadOnlyList<(double, double)> impactFactors = Array.Empty<(double, double)>();

            double z = Random.NextDouble();
            double waterLevel = Random.NextDouble();
            double waveHeightHm0 = Random.NextDouble();
            double fatigueAlpha = Random.NextDouble();
            double fatigueBeta = Random.NextDouble();
            double impactNumberC = Random.NextDouble();

            // Call
            var input = new AsphaltRevetmentWaveImpactFunctionsInput(logFailureTension, averageNumberOfWaves, maximumPeakStress,
                                                                     stiffnessRelation, computationalThickness, outerSlope,
                                                                     widthFactors, depthFactors, impactFactors, z, waterLevel,
                                                                     waveHeightHm0, fatigueAlpha, fatigueBeta, impactNumberC);

            // Assert
            Assert.AreEqual(logFailureTension, input.LogFailureTension);
            Assert.AreEqual(averageNumberOfWaves, input.AverageNumberOfWaves);
            Assert.AreEqual(maximumPeakStress, input.MaximumPeakStress);
            Assert.AreEqual(stiffnessRelation, input.StiffnessRelation);
            Assert.AreEqual(computationalThickness, input.ComputationalThickness);
            Assert.AreEqual(outerSlope, input.OuterSlope);

            Assert.AreSame(widthFactors, input.WidthFactors);
            Assert.AreSame(depthFactors, input.DepthFactors);
            Assert.AreSame(impactFactors, input.ImpactFactors);

            Assert.AreEqual(z, input.Z);
            Assert.AreEqual(waterLevel, input.WaterLevel);
            Assert.AreEqual(waveHeightHm0, input.WaveHeightHm0);
            Assert.AreEqual(fatigueAlpha, input.FatigueAlpha);
            Assert.AreEqual(fatigueBeta, input.FatigueBeta);
            Assert.AreEqual(impactNumberC, input.ImpactNumberC);
        }
    }
}